### PR TITLE
ReduceFnRunner: test when watermark leapfrogs EOW and GC

### PR DESCRIPTION
**Pull request by kennknowles**
This is known to fail in older versions; forward porting regression test.

### CODE REVIEW

File: runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java [hunk 1]
```diff
@@ -246,6 +247,24 @@ public void testOnElementBufferingAccumulating() throws Exception {
     tester.assertHasOnlyGlobalAndFinishedSetsFor(firstWindow);
   }
 
+  @Test
+  public void testTimerGrownSession() throws Exception {
```
Comment (bjchambers) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: This no longer deals with "growing" sessions.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java [hunk 2]
```diff
@@ -246,6 +247,24 @@ public void testOnElementBufferingAccumulating() throws Exception {
...
+    ReduceFnTester<Integer, Iterable<Integer>, IntervalWindow> tester =
+        ReduceFnTester.nonCombining(
+            Sessions.withGapDuration(Duration.millis(10)),
```
Comment (bjchambers) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: May want the equivalent for fixed windows?

File: runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java [hunk 3]
```diff
@@ -246,6 +247,24 @@ public void testOnElementBufferingAccumulating() throws Exception {
...
+            AccumulationMode.ACCUMULATING_FIRED_PANES,
+            Duration.millis(50),
+            ClosingBehavior.FIRE_ALWAYS);
+
+    tester.setAutoAdvanceOutputWatermark(true);
+
+    tester.advanceInputWatermark(new Instant(0));
+    injectElement(tester, 1);
+
+    tester.advanceInputWatermark(new Instant(100));
```
Comment (bjchambers) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: Should indicate what this tests -- specifically, what happens if the input watermark jumps from before EOW to after GC time.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java [hunk 4]
```diff
@@ -246,6 +247,24 @@ public void testOnElementBufferingAccumulating() throws Exception {
...
+  }
```
Comment (bjchambers) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: Should add expectations on what the proper/expected behavior is.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java (hunk 1 above)
Comment (kennknowles) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: Renamed

File: runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java (hunk 2 above)
Comment (kennknowles) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: Addeed

File: runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java (hunk 3 above)
Comment (kennknowles) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: Done

File: runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java (hunk 4 above)
Comment (kennknowles) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: Added check for the expected output, too.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java [hunk 5]
```diff
@@ -261,8 +265,32 @@ public void testTimerGrownSession() throws Exception {
...
+            Duration.millis(50),
+            ClosingBehavior.FIRE_ALWAYS);
+
+    tester.setAutoAdvanceOutputWatermark(true);
+
+    tester.advanceInputWatermark(new Instant(0));
+    injectElement(tester, 1);
     tester.advanceInputWatermark(new Instant(100));
+
+    assertThat(tester.extractOutput(), contains(isSingleWindowedValue(contains(1), 1, 0, 10)));
```
Comment (kennknowles) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: Should also assert first/final pane

Comment (kennknowles) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: Done. Added one more test, too.

**Pull request merged**